### PR TITLE
fix grafana clusterroles on upgrade and make prometheus upgrade less …

### PR DESCRIPTION
**Pull request by StevenTobin**
…destructive

## Additional Information

https://issues.jboss.org/browse/INTLY-3912

## Verification Steps
This PR only affects upgrade, it doesn't touch any part of the install process.

Upgrade log: https://privatebin-it-iso.int.open.paas.redhat.com/?99a00984601420c4#0N+kChKQmx6ORHapLu65eYt9EASxQcFnUOQ8NMSUBLc=

Cluster installed at 1.5.1 -> upgraded with PR: https://master.intlymon-72b1.open.redhat.com/

### Installation Verification
- Ensure the author of the PR has attached a log of the installation run from his branch to the jira or pr and check that it exited as expected.
- Verify the fresh installation is correct on cluster provided by PR author 
### Upgrade Verification
- After installation verification, notify the PR author to begin an upgrade on their cluster
- Ensure the developer of the PR has attached a log of the upgrade run from his branch to the jira or pr and check that it exited as expected. 
- If possible, look at the tasks that ran and see they match the PR
- Verify the upgrade is correct on cluster provided by PR author 

1. Install release-1.5.1
2. Upgrade from this branch
3. Ensure application-monitoring-operator is 0.0.27
4 Ensure prometheus-operator is 0.33.0
5. Ensure the `- --deny-namespaces=openshift-monitoring"` argument is on the prometheus-operator pod

## Is an upgrade task required and are there additional steps needed to test this?

- [x] Yes
- [ ] No

- [ ] Tested Installation
- [ ] Tested Uninstallation
- [x] Tested or Created follow on for Upgrade

### CODE REVIEW

**Comment by david-martin** [comment]
Log of `` during upgrade:

```
middleware-monitoring   grafana-deployment-54447b6c88-shjz7   2/2   Terminating   0     10m
middleware-monitoring   grafana-operator-574f85cf79-l587b   1/1   Terminating   0     47m
middleware-monitoring   grafana-deployment-54447b6c88-shjz7   0/2   Terminating   0     10m
middleware-monitoring   grafana-operator-574f85cf79-l587b   0/1   Terminating   0     47m
middleware-monitoring   grafana-operator-574f85cf79-l587b   0/1   Terminating   0     47m
middleware-monitoring   grafana-operator-574f85cf79-l587b   0/1   Terminating   0     47m
middleware-monitoring   prometheus-operator-564bc9c55b-lbmqb   1/1   Terminating   0     47m
middleware-monitoring   prometheus-operator-564bc9c55b-lbmqb   0/1   Terminating   0     48m
middleware-monitoring   prometheus-operator-564bc9c55b-lbmqb   0/1   Terminating   0     48m
middleware-monitoring   prometheus-operator-564bc9c55b-lbmqb   0/1   Terminating   0     48m
middleware-monitoring   prometheus-operator-5dff665d54-cz4g2   0/1   Pending   0     0s
middleware-monitoring   prometheus-operator-5dff665d54-cz4g2   0/1   Pending   0     0s
middleware-monitoring   prometheus-operator-5dff665d54-cz4g2   0/1   ContainerCreating   0     0s
middleware-monitoring   application-monitoring-operator-76d8945cfb-hbjk2   1/1   Terminating   0     48m
middleware-monitoring   application-monitoring-operator-76d8945cfb-hbjk2   0/1   Terminating   0     48m
middleware-monitoring   prometheus-operator-5dff665d54-cz4g2   1/1   Running   0     4s
middleware-monitoring   grafana-deployment-54447b6c88-shjz7   0/2   Terminating   0     11m
middleware-monitoring   grafana-deployment-54447b6c88-shjz7   0/2   Terminating   0     11m
middleware-monitoring   application-monitoring-operator-6f87c9c5cb-bsw5r   0/1   Pending   0     0s
middleware-monitoring   application-monitoring-operator-6f87c9c5cb-bsw5r   0/1   Pending   0     0s
middleware-monitoring   application-monitoring-operator-6f87c9c5cb-bsw5r   0/1   ContainerCreating   0     0s
middleware-monitoring   alertmanager-application-monitoring-0   3/3   Terminating   0     48m
middleware-monitoring   alertmanager-application-monitoring-0   3/3   Terminating   0     48m
middleware-monitoring   application-monitoring-operator-76d8945cfb-hbjk2   0/1   Terminating   0     48m
middleware-monitoring   application-monitoring-operator-76d8945cfb-hbjk2   0/1   Terminating   0     48m
middleware-monitoring   application-monitoring-operator-6f87c9c5cb-bsw5r   0/1   Running   0     3s
middleware-monitoring   alertmanager-application-monitoring-0   0/3   Pending   0     0s
middleware-monitoring   alertmanager-application-monitoring-0   0/3   Pending   0     0s
middleware-monitoring   alertmanager-application-monitoring-0   0/3   ContainerCreating   0     1s
middleware-monitoring   grafana-operator-574f85cf79-ndw7f   0/1   Pending   0     0s
middleware-monitoring   grafana-operator-574f85cf79-ndw7f   0/1   Pending   0     0s
middleware-monitoring   grafana-operator-574f85cf79-ndw7f   0/1   ContainerCreating   0     1s
middleware-monitoring   grafana-deployment-bb9d86548-trss8   0/2   Pending   0     0s
middleware-monitoring   grafana-deployment-bb9d86548-trss8   0/2   Pending   0     0s
middleware-monitoring   grafana-deployment-bb9d86548-trss8   0/2   Init:0/1   0     0s
middleware-monitoring   alertmanager-application-monitoring-0   3/3   Running   0     8s
middleware-monitoring   prometheus-application-monitoring-0   5/5   Terminating   1     10m
middleware-monitoring   grafana-operator-574f85cf79-ndw7f   0/1   Running   0     8s
middleware-monitoring   grafana-deployment-bb9d86548-trss8   0/2   PodInitializing   0     3s
middleware-monitoring   prometheus-application-monitoring-0   0/5   Terminating   1     10m
middleware-monitoring   application-monitoring-operator-6f87c9c5cb-bsw5r   1/1   Running   0     15s
middleware-monitoring   grafana-deployment-bb9d86548-trss8   2/2   Running   0     4s
middleware-monitoring   grafana-operator-574f85cf79-ndw7f   1/1   Running   0     14s
middleware-monitoring   grafana-deployment-bb9d86548-trss8   2/2   Terminating   0     15s
middleware-monitoring   grafana-deployment-bb9d86548-gls7j   0/2   Pending   0     1s
middleware-monitoring   grafana-deployment-bb9d86548-gls7j   0/2   Pending   0     1s
middleware-monitoring   grafana-deployment-bb9d86548-gls7j   0/2   Init:0/1   0     1s
middleware-monitoring   grafana-deployment-bb9d86548-trss8   0/2   Terminating   0     15s
middleware-monitoring   grafana-deployment-bb9d86548-trss8   0/2   Terminating   0     17s
middleware-monitoring   grafana-deployment-bb9d86548-trss8   0/2   Terminating   0     17s
middleware-monitoring   grafana-deployment-bb9d86548-gls7j   0/2   Init:0/1   0     4s
middleware-monitoring   grafana-deployment-bb9d86548-gls7j   0/2   PodInitializing   0     6s
middleware-monitoring   grafana-deployment-bb9d86548-gls7j   2/2   Running   0     7s
middleware-monitoring   prometheus-application-monitoring-0   0/5   Terminating   1     11m
middleware-monitoring   prometheus-application-monitoring-0   0/5   Terminating   1     11m
middleware-monitoring   prometheus-application-monitoring-0   0/5   Pending   0     0s
middleware-monitoring   prometheus-application-monitoring-0   0/5   Pending   0     0s
middleware-monitoring   prometheus-application-monitoring-0   0/5   ContainerCreating   0     0s
sso   keycloak-operator-857d95786c-wll5j   0/1   Pending   0     0s
sso   keycloak-operator-857d95786c-wll5j   0/1   Pending   0     0s
sso   keycloak-operator-857d95786c-wll5j   0/1   ContainerCreating   0     0s
user-sso   keycloak-operator-9ff9dc64c-rwqd9   0/1   Pending   0     0s
user-sso   keycloak-operator-9ff9dc64c-rwqd9   0/1   Pending   0     0s
user-sso   keycloak-operator-9ff9dc64c-rwqd9   0/1   ContainerCreating   0     1s
sso   keycloak-operator-857d95786c-wll5j   1/1   Running   0     3s
middleware-monitoring   prometheus-application-monitoring-0   4/5   Running   0     6s
sso   keycloak-operator-855478547f-lwr8d   1/1   Terminating   0     46m
middleware-monitoring   prometheus-application-monitoring-0   5/5   Running   0     9s
sso   keycloak-operator-855478547f-lwr8d   0/1   Terminating   0     46m
sso   keycloak-operator-855478547f-lwr8d   0/1   Terminating   0     46m
sso   keycloak-operator-855478547f-lwr8d   0/1   Terminating   0     46m
user-sso   keycloak-operator-9ff9dc64c-rwqd9   1/1   Running   0     10s
user-sso   keycloak-operator-6745846967-j7k5w   1/1   Terminating   0     12m
user-sso   keycloak-operator-6745846967-j7k5w   0/1   Terminating   0     12m
user-sso   keycloak-operator-6745846967-j7k5w   0/1   Terminating   0     12m
user-sso   keycloak-operator-6745846967-j7k5w   0/1   Terminating   0     12m
middleware-monitoring   grafana-deployment-bb9d86548-gls7j   2/2   Terminating   0     31s
middleware-monitoring   grafana-deployment-54447b6c88-qn24n   0/2   Pending   0     0s
middleware-monitoring   grafana-deployment-54447b6c88-qn24n   0/2   Pending   0     0s
middleware-monitoring   grafana-deployment-54447b6c88-qn24n   0/2   Init:0/1   0     0s
middleware-monitoring   grafana-deployment-bb9d86548-gls7j   0/2   Terminating   0     32s
middleware-monitoring   grafana-deployment-54447b6c88-qn24n   0/2   Init:0/1   0     2s
middleware-monitoring   grafana-deployment-54447b6c88-qn24n   0/2   PodInitializing   0     5s
middleware-monitoring   grafana-deployment-bb9d86548-gls7j   0/2   Terminating   0     38s
middleware-monitoring   grafana-deployment-bb9d86548-gls7j   0/2   Terminating   0     38s
middleware-monitoring   grafana-deployment-54447b6c88-qn24n   2/2   Running   0     9s
```

After upgrade, previous metrics still visible in grafana, so prometheus storage has not been affected.

image

Version of components after:
```
oc get deployment application-monitoring-operator -n middleware-monitoring -o yaml | grep image:
        image: quay.io/integreatly/application-monitoring-operator:0.0.27

oc get deployment prometheus-operator -n middleware-monitoring -o yaml | grep image:
        image: quay.io/coreos/prometheus-operator:v0.33.0
```

No errors in grafana operator after:
```
oc -n middleware-monitoring logs grafana-operator-574f85cf79-ndw7f| tail -20
{"level":"info","ts":1572628488.5214474,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628488.7246633,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628488.9057171,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628489.3421123,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628489.950081,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628490.558013,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628491.1801388,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628491.7524674,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628492.3722985,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628492.944421,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628493.5910966,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628494.177436,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628494.7751591,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628495.3574717,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628495.996271,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628496.5759757,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628497.173613,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628497.7670305,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628498.3722863,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
{"level":"info","ts":1572628498.9715295,"logger":"controller_grafanadashboard","caller":"grafanadashboard/dashboard_controller.go:147","msg":"dashboard reconciled but no changes"}
```

**Comment by david-martin** [comment]
@pb82  Can you ack this when you have a moment?

File: roles/middleware_monitoring/tasks/upgrade/prometheus.yml [hunk 1]
```diff
@@ -5,39 +5,35 @@
...
-  register: get_statefulset_cmd
-  failed_when: get_statefulset_cmd.rc == 0
-  changed_when: "'NotFound' in get_statefulset_cmd.stderr"
-  retries: 10
-  delay: 5
+- name: Add the deny-namespaces argument
+  lineinfile:
+    path: /tmp/prometheus-operator.yml
+    insertafter: "- --prometheus-config-reloader="
+    line: "        - --deny-namespaces=openshift-monitoring"
```
Comment (pb82) [inline roles/middleware_monitoring/tasks/upgrade/prometheus.yml]: @david-martin @StevenTobin On master there are more entries after `deny-namespaces`: https://github.com/integr8ly/application-monitoring-operator/blob/master/templates/prometheus-operator.yaml#L25

Is it ok to only deny openshift-monitoring in 1.5.2?

Comment (david-martin) [reply roles/middleware_monitoring/tasks/upgrade/prometheus.yml]: Yes, given that v 0.0.27 or the application monitoring operator is being used, this matches what a fresh install will do.
https://github.com/integr8ly/application-monitoring-operator/blob/0.0.27/templates/prometheus-operator.yaml#L25

master will use a later version that brings in that change.

**Comment by david-martin** [comment]
@pb82 can you add a /lgtm if you're good to merge

**Pull request merged**